### ISSUE TITLE
does not pass args capability to geckodriver

### ISSUE THREAD

**Issue by stuart-little** [SeleniumHQ/selenium#9472]
## 🐛 Bug Report

I am unable to append options to the `geckodriver` [args capability](https://developer.mozilla.org/en-US/docs/Web/WebDriver/Capabilities/firefoxOptions) via the [desired_capabilities](https://selenium-python.readthedocs.io/api.html#module-selenium.webdriver.firefox.webdriver) parameter.

## To Reproduce

I am running this:
```
#!/usr/bin/env python

from selenium.webdriver import Firefox, DesiredCapabilities
import time

capabilities = DesiredCapabilities.FIREFOX.copy()
capabilities['args'] = ["-headless"]

print(capabilities)

webdriver = Firefox(desired_capabilities=capabilities)
webdriver.get("https://google.com")

while 1:
    time.sleep(0.1)
```

## Expected behavior

A *headless* instance visiting `google.com`, as I gather from the [docs](https://developer.mozilla.org/en-US/docs/Web/WebDriver/Capabilities/firefoxOptions): the `-headless` string being one of the members of the `args` array should mandate that headlessness.

## Actual behavior

A *visible* window. 

## Test script or set of commands reproducing this issue

Provided above. 

## Environment

OS: `Linux x86_64`
Browser: `Firefox`
Browser version: `Mozilla Firefox 88.0.1`
Browser Driver version: `geckodriver 0.29.0`
Language Bindings version: `Python`'s `selenium 3.141.0` [package](https://pypi.org/project/selenium/)

## Additional info
I have that other types of `args` are not passed either. For instance, `capabilities['args'] = ["-profile", "<PATH-TO-EXISTING-PROFILE>"]` does not result in `geckodriver` using that existing profile, as suggested by [the docs](https://developer.mozilla.org/en-US/docs/Web/WebDriver/Capabilities/firefoxOptions) that it should.

**Comment by AutomatedTester** [comment]
Please use the `FirefoxOptions` class as you're not passing in the capabilities properly. The options class simplifies that for you.

**Comment by stuart-little** [comment]
Thank you @AutomatedTester  (though a pointer to the docs could have been a bit more explicit..). I am now doing this, for future reference, in case anyone finds their way here:
```
from selenium.webdriver.firefox.options import Options
import time

options = Options()
options.add_argument("-profile")
options.add_argument("<PATH-TO-PROFILE>")

webdriver = Firefox(desired_capabilities=options.to_capabilities())
webdriver.get("https://google.com")

while 1:
    time.sleep(0.1)
```
This [fails differently](https://github.com/SeleniumHQ/selenium/issues/9474#issue-890987338), but at least `geckodriver.log` sees my `-profile` argument:
```
$ cat geckodriver.log
---
1620906396749   geckodriver     INFO    Listening on 127.0.0.1:49235
1620906396752   mozrunner::runner       INFO    Running command: "/usr/bin/firefox" "--marionette" "-profile" "<PATH-TO-PROFILE>" "-foreground" "-no-remote"
```

**Comment by AutomatedTester** [comment]
I don't recommend passing around profiles that way to be clear.

The [API docs](https://www.selenium.dev/selenium/docs/api/py/webdriver_firefox/selenium.webdriver.firefox.options.html#module-selenium.webdriver.firefox.options) has other mechanisms. If you want a profile I recommend that you use `set_preference` and if you need install addons then you can use [`install_addon`](https://www.selenium.dev/selenium/docs/api/py/webdriver_firefox/selenium.webdriver.firefox.webdriver.html#selenium.webdriver.firefox.webdriver.WebDriver.install_addon). That will allow geckodriver to build a nice clean profile and is scalable.

**Comment by stuart-little** [comment]
I understand it's not advisable, but I need a way to store whatever the profile gathers (cookies, etc.). I know about [get_cookies()](https://selenium-python.readthedocs.io/api.html#selenium.webdriver.remote.webdriver.WebDriver.get_cookies) but in this case they are produced by an iframe, and I [cannot retrieve](https://bugzilla.mozilla.org/show_bug.cgi?id=1427807) those.

So: I need a way to start up a browser, have it run a bunch of tasks automatically, and then keep access to *everything* that was retrieved (the entirety of that profile). I thought an existing profile would be the best way to gain that persistence. 

I can also quit the driver with a `SIGINT` and have it leave behind the temporary profile, but that's [regarded as](https://github.com/mozilla/geckodriver/issues/299) a bug.

**Comment by stuart-little** [comment]
@AutomatedTester Incidentally, I'm [aware](https://selenium-python.readthedocs.io/api.html#module-selenium.webdriver.firefox.webdriver) that one can pass `firefox_profile=<whatever>` or `options.profile=<whatever>`, but this does not do wat I want: I have checked that both create *copies* of that profile to a temporary location. If I end the driver cleanly whatever they've gathered is wiped out. 

For instance, if I delete the `cookies.sqlite` from `<PROFILE>` and then pass `firefox_profile=<PROFILE>` when building the driver, it will *not* leave behind a `cookies.sqlite` file in my original profile directory. 

So: what I am after is a way to use a profile *persistently* with `geckodriver`. To save my life I cannot do this.

**Comment by AutomatedTester** [comment]
> @AutomatedTester Incidentally, I'm [aware](https://selenium-python.readthedocs.io/api.html#module-selenium.webdriver.firefox.webdriver) that one can pass `firefox_profile=<whatever>` or `options.profile=<whatever>`, but this does not do wat I want: I have checked that both create _copies_ of that profile to a temporary location. If I end the driver cleanly whatever they've gathered is wiped out.
> 
> For instance, if I delete the `cookies.sqlite` from `<PROFILE>` and then pass `firefox_profile=<PROFILE>` when building the driver, it will _not_ leave behind a `cookies.sqlite` file in my original profile directory.
> 
> So: what I am after is a way to use a profile _persistently_ with `geckodriver`. To save my life I cannot do this.

What you are after is going against how Selenium, and Geckodriver, do things in general. Selenium originally, and then geckodriver too, copy the profile so that if you pass it in that you get the same start point as always. We found that it was a major footgun to allow things to persist indefinitely as well as being a security issue if people share these profiles around.

If you want to get the resultant profile after running your automation code I suggest you do something like

```python
from selenium.webdriver.firefox.options import Options
from selenium import webdriver
options = Options()
options.profile = "/path/to/profile"
driver = webdriver.Firefox(options=options)
# Do whatever you want
profile = driver.firefox_profile
# copy the profile somewhere
driver.quit()
```

The thing you're trying to do is not how Selenium/Geckodriver was designed so you're going to have to try be creative here.

**Comment by stuart-little** [comment]
@AutomatedTester Thank you: confirming that `Selenium` does not make this easy by default is at least a sanity check of sorts. I thought it was me missing some easy persistence switch.

As to being creative, yes, I figured I'd have to cobble something together.. What I've settled on for now, since I'm specifically interested in cookies, is hard-linking the temp profile's `cookies.sqlite` file to a location of my choosing. Then, when it exits, writes the cookies to that database, releases the lock and then deletes it, I have the copy.

Anyway, I think I can work with this. Thank you very much for all the help and the time you invested in this exchange.
